File: src/main/java/com/esprit/stage/Service/StatisticsService.java
package com.esprit.stage.Service;

import com.esprit.stage.Repository.StatisticsRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class StatisticsService {

    @Autowired
    private StatisticsRepository statisticsRepository;

    // 📊 Nombre d’offres par entreprise (global)
    public Map<String, Long> getOffersByCompany() {
        return convertToMap(statisticsRepository.countOffersByCompany());
    }

    // 📈 Nombre d’offres créées par jour (global)
    public Map<String, Long> getOffersByDay() {
        return convertToMap(statisticsRepository.countOffersByDay());
    }
    public Double getSubmissionToOfferRatio() {
        Double ratio = statisticsRepository.getSubmissionToOfferRatio();
        return ratio != null ? ratio : 0.0;  // ✅ Assurer une valeur par défaut
    }



    public Map<String, Long> getOffersByLocation() {
        List<Object[]> results = statisticsRepository.countOffersByLocation();
        return results.stream().collect(Collectors.toMap(
                data -> String.valueOf(data[0]),  // Nom de la ville
                data -> (Long) data[1]           // Nombre d'offres
        ));
    }


    // 📊 Nombre de candidatures par offre (global)
    public Map<String, Long> getApplicationsByOffer() {
        return convertToMap(statisticsRepository.countApplicationsByOffer());
    }

    // 📊 Nombre de candidatures par entreprise (global)
    public Map<String, Long> getApplicationsByCompany() {
        return convertToMap(statisticsRepository.countApplicationsByCompany());
    }

    // 📊 Nombre d’offres publiées pour une entreprise spécifique (RH)
    public Long getOffersByCompanyRH(String companyName) {
        return statisticsRepository.countOffersByCompanyRH(companyName);
    }

    // 📈 Nombre d’offres créées par jour pour une entreprise spécifique (RH)
    public Map<String, Long> getOffersByCompanyPerDay(String companyName) {
        return convertToMap(statisticsRepository.countOffersByCompanyPerDay(companyName));
    }

    // 📊 Nombre de candidatures reçues par une entreprise spécifique (RH)
    public Long getApplicationsByCompanyRH(String companyName) {
        return statisticsRepository.countApplicationsByCompanyRH(companyName);
    }

    // 🔢 Ratio candidatures/offres pour une entreprise spécifique (RH)
    public Double getSubmissionToOfferRatioByCompany(String companyName) {
        Double ratio = statisticsRepository.getSubmissionToOfferRatioByCompany(companyName);
        return ratio != null ? ratio : 0.0;
    }

    // 📌 Conversion des résultats en JSON
    private Map<String, Long> convertToMap(List<Object[]> results) {
        return results.stream().collect(Collectors.toMap(
                data -> String.valueOf(data[0]),
                data -> (Long) data[1]
        ));
    }
}

File: piProject-nourAngular/piProject-nourAngular/src/app/services/statistics.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({ providedIn: 'root' })

export class StatisticsService {
  
  private baseUrl = 'http://localhost:8093/api/internship-offers/statistics';

  constructor(private http: HttpClient) {}

  // 📊 Nombre d’offres par entreprise (Admin)
  getOffersByCompany(): Observable<{ [key: string]: number }> {
    return this.http.get<{ [key: string]: number }>(`${this.baseUrl}/global/offers-by-company`);
  }

  // 📈 Nombre d’offres créées par jour (Admin)
  getOffersByDay(): Observable<{ [key: string]: number }> {
    return this.http.get<{ [key: string]: number }>(`${this.baseUrl}/global/offers-by-day`);
  }

  // 📍 Répartition des offres par ville (Admin)
  getOffersByLocation(): Observable<{ [key: string]: number }> {
    return this.http.get<{ [key: string]: number }>(`${this.baseUrl}/global/offers-by-location`);
  }

  // 🎯 Répartition des types de stages (PFE vs Summer) (Admin)
  getOffersByType(): Observable<{ [key: string]: number }> {
    return this.http.get<{ [key: string]: number }>(`${this.baseUrl}/global/offers-by-type`);
  }

  // 📌 Statut des offres (Open/Closed/Archived) (Admin)
  getOffersByStatus(): Observable<{ [key: string]: number }> {
    return this.http.get<{ [key: string]: number }>(`${this.baseUrl}/global/offers-by-status`);
  }

  // 📊 Nombre de candidatures par offre (Admin)
  getApplicationsByOffer(): Observable<{ [key: string]: number }> {
    return this.http.get<{ [key: string]: number }>(`${this.baseUrl}/global/applications-by-offer`);
  }

  // 🏢 Nombre de candidatures par entreprise (Admin)
  getApplicationsByCompany(): Observable<{ [key: string]: number }> {
    return this.http.get<{ [key: string]: number }>(`${this.baseUrl}/global/applications-by-company`);
  }

  // 🌐 Technologies les plus mentionnées (Admin - Word Cloud)
  getMostUsedTechnologies(): Observable<{ [key: string]: number }> {
    return this.http.get<{ [key: string]: number }>(`${this.baseUrl}/global/technologies`);
  }

  // 🔥 Ratio candidatures/offres (Admin)
  getSubmissionToOfferRatio(): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/global/submission-to-offer-ratio`);
  }

  // 📊 Nombre total d’offres publiées pour une entreprise spécifique (RH)
  getOffersByCompanyRH(companyName: string): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/company/${companyName}/offers`);
  }

  // 📈 Nombre d’offres créées par jour pour une entreprise spécifique (RH)
  getOffersByCompanyPerDay(companyName: string): Observable<{ [key: string]: number }> {
    return this.http.get<{ [key: string]: number }>(`${this.baseUrl}/company/${companyName}/offers-by-day`);
  }

  // 📊 Nombre de candidatures reçues pour une entreprise spécifique (RH)
  getApplicationsByCompanyRH(companyName: string): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/company/${companyName}/applications`);
  }

  // 🔢 Ratio candidatures/offres pour une entreprise spécifique (RH)
  getSubmissionToOfferRatioByCompany(companyName: string): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/company/${companyName}/submission-to-offer-ratio`);
  }
}

